src/platform/storage/zephyr/nvs/storage.c

```diff
@@ -163,9 +197,149 @@ mender_storage_init(void) {
                      mender_storage_nvs_handle.sector_count,
                      (mender_storage_nvs_handle.sector_count - 1) * mender_storage_nvs_handle.sector_size);
 
+#ifdef CONFIG_MENDER_DEPLOYMENT_LOGS
+    /* Initialize the Flash Circular Buffer (FCB) for deployment logs. */
+
+    /* just some fixed magic value the FCB implementation uses to mark
+       in-use/erased/... sectors */
+    depl_logs_buffer.f_magic       = 0x2AABEE35;
+    depl_logs_buffer.f_version     = 0; /* we don't version the data so 0 always */
+    depl_logs_buffer.f_sector_cnt  = CONFIG_MENDER_STORAGE_DEPLOYMENT_LOGS_SECTORS;
+    depl_logs_buffer.f_scratch_cnt = 0; /* no scratch sector, we don't use it */
+    depl_logs_buffer.f_sectors     = flash_sectors + CONFIG_MENDER_STORAGE_NVS_SECTOR_COUNT;
+
+    if (0 != (result = fcb_init(FIXED_PARTITION_ID(MENDER_STORAGE_LABEL), &depl_logs_buffer))) {
+        mender_log_debug("Failed to initialize deployment logs FCB, erasing the particular flash area");
+
+        const struct flash_area *fap;
+        if (0 != (result = flash_area_open(FIXED_PARTITION_ID(MENDER_STORAGE_LABEL), &fap))) {
+            mender_log_error("Unable to open the Mender storage flash area");
+            return MENDER_FAIL;
+        }
+
+        /* flatten means: Erase flash area or fill with erase-value. */
+        result = flash_area_flatten(
+            fap, depl_logs_buffer.f_sectors[0].fs_off, depl_logs_buffer.f_sectors[0].fs_off * CONFIG_MENDER_STORAGE_DEPLOYMENT_LOGS_SECTORS);
+        flash_area_close(fap);
+        if (0 != result) {
+            mender_log_error("Failed to erase the flash area for deployment logs");
+        }
+        /* Now, try again. */
+        if (0 != (result = fcb_init(FIXED_PARTITION_ID(MENDER_STORAGE_LABEL), &depl_logs_buffer))) {
+            mender_log_error("Unable to initialize the Flash Circular Buffer for deployment logs [%d]", -result);
+            return MENDER_FAIL;
+        }
+    }
+    mender_log_debug("Initialized deployment logs FCB with %u sectors (%zd bytes available)",
+                     depl_logs_buffer.f_sector_cnt,
+                     (depl_logs_buffer.f_sector_cnt - 1) * mender_storage_nvs_handle.sector_size);
+#endif /* CONFIG_MENDER_DEPLOYMENT_LOGS */
+
     return MENDER_OK;
 }
 
+#ifdef CONFIG_MENDER_DEPLOYMENT_LOGS
+mender_err_t
+mender_storage_deployment_log_append(const char *msg, size_t msg_size) {
+    int          result;
+    mender_err_t ret = MENDER_OK;
+
+    msg_size = MIN(DEPL_LOGS_MAX_MSG_LEN + 1, msg_size);
+
+    const struct flash_area *fap;
+    if (0 != (result = flash_area_open(FIXED_PARTITION_ID(MENDER_STORAGE_LABEL), &fap))) {
+        mender_log_error("Unable to open the Mender storage flash area");
+        return MENDER_FAIL;
+    }
+
+    struct fcb_entry entry = { 0 };
+    result                 = fcb_append(&depl_logs_buffer, msg_size, &entry);
+    if (-ENOSPC == result) {
```

Comment:
All this handling could be done on a separate static function. It would just read better :eyes: